### ISSUE TITLE
ADCE deletes label, but leaves a branch to the label.

### ISSUE THREAD

**Issue by s-perron** [KhronosGroup/SPIRV-Tools#1075]
If an empty structured-if has a merge block that is also the continue target of a loop construct, the block in the if is removed, but the branch is left as is.  This leaves a use of a label id that has no definition.

The following SPIR-V demonstrates the problem.  Assemble it, and run it through ADCE.

```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %2 "main" %3
               OpExecutionMode %2 OriginUpperLeft
               OpSource GLSL 430
               OpSourceExtension "GL_GOOGLE_cpp_style_line_directive"
               OpSourceExtension "GL_GOOGLE_include_directive"
               OpName %2 "main"
               OpDecorate %3 Location 0
          %4 = OpTypeVoid
          %5 = OpTypeFunction %4
          %6 = OpTypeFloat 32
          %8 = OpConstant %6 0
          %9 = OpTypeInt 32 1
         %10 = OpTypePointer Function %9
         %11 = OpConstant %9 0
         %12 = OpConstant %9 10
         %13 = OpTypeBool
         %14 = OpConstant %6 1
         %15 = OpConstant %9 1
         %16 = OpTypePointer Output %6
          %3 = OpVariable %16 Output
          %2 = OpFunction %4 None %5
         %20 = OpLabel
               OpBranch %21
         %21 = OpLabel
         %22 = OpPhi %6 %8 %20 %23 %24
         %25 = OpPhi %9 %11 %20 %26 %24
               OpLoopMerge %27 %24 None
               OpBranch %28
         %28 = OpLabel
         %29 = OpSLessThan %13 %25 %12
               OpBranchConditional %29 %30 %27
         %30 = OpLabel
         %23 = OpFAdd %6 %22 %14
         %31 = OpConvertFToS %9 %23
         %32 = OpSGreaterThan %13 %25 %31
               OpSelectionMerge %24 None
               OpBranchConditional %32 %33 %24
         %33 = OpLabel
               OpBranch %24
         %24 = OpLabel
         %26 = OpIAdd %9 %25 %15
               OpBranch %21
         %27 = OpLabel
               OpStore %3 %22
               OpReturn
               OpFunctionEnd
```

**Comment by greg-lunarg** [comment]
Good catch! 

A related question: how do you interpret the following from the SPIR-V spec, giving requirements for the Shader capability:

Selections must be structured, having an OpSelectionMerge instruction in their header.

Specifically, how do you interpret the word "Selections"? Specifically, does this mean that every OpBranchConditional must be preceded by an OpSelectionMerge? If not, which ones must be?

**Comment by s-perron** [comment]
I believe every OpBranchConditional must be related to an Op*Merge instruction.  The OpBranchConditional in BB 28 does not have an OpSelectionMerge because it is the loop exit related to the OpLoopMerge in BB 21.

The example above was generated by `glslc` from the shaderc repository, with a small change by hand.  The original source was

```
#version 430

layout(std430) buffer U_t { float g_F[10]; };

layout(location = 0) out float o;

void main(void) {
  float s = 0.0;
  for (int j = 0; j < 10; j++) {
    s += 1.0;
    if (j > int(s));
  }
  o = s;
}
```

**Comment by dneto0** [comment]
#1088 has been merged.

**Comment by dneto0** [comment]
> Specifically, how do you interpret the word "Selections"? Specifically, does this mean that every OpBranchConditional must be preceded by an OpSelectionMerge? If not, which ones must be?

This unfortunately vague, and has been a topic of debate.

No, not every OpBranchConditional has to have a merge instruction.  Question is, which ones.

I'd like to say "Selections" should be interpreted as a "selection construct" but that's defined in terms of a "selection header" which itself is not defined in the spec.  You might guess that it means "a header block which is not a loop header".  But then that would mean "a block that has an OpSelectionMerge instruction".  Which then leads to a circularity.  (Ouch.)

I assert the way to think about it is in terms of dominating header and merge blocks:
* Consider a branch instruction in block B, branching to block C. (For a conditional branch, C will stand in the text below for either of the two targets.)
* Intution: You need a merge instruction in B if its branch introduces a "new" divergence in control flow.
* More formally: You need a merge instruction in B if all of the following are true for C:
   * C is *not* the merge or continue target for an enclosing header block H.  (H encloses B if B is strictly dominated by H, and B is not dominated by the merge block named by H.)
   * C is *not* one of the case targets of an OpSwitch

I think I got that right.

Basically, branches to targets "out" of a construct are not counted as forcing the need for a selection merge.

**Comment by greg-lunarg** [comment]
Thanks David, that's very helpful. Perhaps that statement in the spec could be amended to?:

Selections must be structured, having an OpSelectionMerge instruction in their header, unless they are an allowed branch out of the enclosing control flow construct.
